erts/emulator/beam/erl_debugger.c

```diff
@@ -449,6 +449,173 @@ erts_internal_notify_breakpoint_hit_3(BIF_ALIST_3) {
+    }
+}
+
+BIF_RETTYPE
+erl_debugger_breakpoints_3(BIF_ALIST_3) {
+    Eterm module_name, fun_name, arity, badkey_source, *hp;
+    Uint arity_val;
+    Module* modp;
+    ErtsCodeIndex code_ix = erts_active_code_ix();
+    const BeamCodeHeader* code_hdr;
+
+    BIF_UNDEF_IF_NO_DEBUGGER_SUPPORT();
+
+    module_name = BIF_ARG_1;
+    if (is_not_atom(module_name)) {
+        BIF_ERROR(BIF_P, BADARG);
+    }
+
+    fun_name = BIF_ARG_2;
+    if (is_not_atom(fun_name)) {
+        BIF_ERROR(BIF_P, BADARG);
+    }
+
+    arity = BIF_ARG_3;
+    if (is_not_small(BIF_ARG_3)) {
+        BIF_ERROR(BIF_P, BADARG);
+    }
+    arity_val = unsigned_val(BIF_ARG_3);
+
+    modp = erts_get_module(module_name, code_ix);
+    if (modp == NULL) {
+        badkey_source = module_name;
+        goto error;
+    }
+
+    code_hdr = modp->curr.code_hdr;
+    if (code_hdr == NULL) {
+        badkey_source = module_name;
+        goto error;
+    }
+
+    for(int fun_idx = 0; fun_idx < code_hdr->num_functions; fun_idx++) {
+        const ErtsCodeInfo *fun_info = code_hdr->functions[fun_idx];
+        const ErtsCodeMFA *mfa = &fun_info->mfa;
+        if (mfa->function == fun_name && mfa->arity == arity_val) {
+            const BeamCodeLineTab *lt = code_hdr->line_table;
+            Eterm result = build_breakpoints_info_for_fun(BIF_P, lt, fun_idx);
+            Eterm *hp = HAlloc(BIF_P, 3);
+            BIF_RET(TUPLE2(hp, am_ok, result));
+        }
+    }
+
+    hp = HAlloc(BIF_P, 3);
+    badkey_source = TUPLE2(hp, fun_name, arity);
+    goto error;
+
+    {
+        Eterm *hp1, *hp2;
+    error:
+
+        hp1 = HAlloc(BIF_P, 6);
+        hp2 = hp1 + 3;
+        BIF_RET(TUPLE2(hp2, am_error, TUPLE2(hp1, am_badkey, badkey_source)));
+    }
+}
+
+static Eterm build_breakpoints_info_for_fun(Process *c_p, const BeamCodeLineTab *lt, int fun_idx) {
+    Eterm *hp = HAlloc(c_p, MAP0_SZ);
+    Eterm fun_lines = MAP0(hp);
+
+    const void **fun_line_starts = lt->func_tab[fun_idx];
+    const void *next_fun_start = lt->func_tab[fun_idx+1][0];
+    int loc_tab_idx = lt->func_tab[fun_idx] - lt->func_tab[0];
+
+    for(int fun_line=0; fun_line_starts[fun_line] < next_fun_start; fun_line++) {
+        Uint32 loc;
+        Eterm line, status;
+        const Eterm *prev_status;
+
+        if (lt->loc_size == 2) {
+            loc = lt->loc_tab.p2[loc_tab_idx + fun_line];
+        } else {
+            ASSERT(lt->loc_size == 4);
+            loc = lt->loc_tab.p4[loc_tab_idx + fun_line];
+        }
+        if (loc == LINE_INVALID_LOCATION) {
+            continue;
+        }
+
+        line = make_small(LOC_LINE(loc));
+
+        switch (erts_is_line_breakpoint_code(fun_line_starts[fun_line])) {
+        case IS_NOT_LINE_BP:
+            continue;
+        case IS_ENABLED_LINE_BP:
+            status = am_true;
+            break;
+        case IS_DISABLED_LINE_BP:
+            status = am_false;
+            break;
+        default:
+            ASSERT(0);
+            continue;
```

Comment:
Nitpick, all `enum` values are handled already and the compiler should know that:

```suggestion
```

Reply:
> the compiler should know that

True but it currently emits a *warning* if I comment out the `default` and and additional case, not an error. At least for me, I get lots of warnings when building the VM (mostly `-Warray-bounds`, I think because `foo[1]` is used throughout instead of `foo[]` for "flexible arrays"), so easy to miss. Maybe `-Wswitch` should be treated as error?